cpp/src/rolling/rolling.cu

```diff
@@ -0,0 +1,260 @@
+
+    // dynamic window handling
+    if (window_col != NULL) window = window_col[i];
+    if (min_periods_col != NULL) min_periods = max(min_periods_col[i], 1);	// at least one observation is required
+    if (forward_window_col != NULL) forward_window = forward_window_col[i];
+
+    // compute bounds
+    gdf_size_type start_index = max((gdf_size_type)0, i - window + 1);
+    gdf_size_type end_index = min(nrows, i + forward_window + 1);       // exclusive
+
+    // aggregate
+    for (size_t j = start_index; j < end_index; j++) {
+      bool const input_is_valid{gdf_is_valid(in_col_valid, j)};
+      if (input_is_valid) {
+        val = op(in_col[j], val);
+        count++;
+      }
+    }
+
+    // check if we have enough input samples
+    bool output_is_valid = (count >= min_periods);
+
+    // set the mask
+    MaskType const result_mask{__ballot_sync(active_threads, output_is_valid)};
+    gdf_index_type const out_mask_location = i / BITS_PER_MASK;
+    MaskType* const __restrict__ out_mask32 = reinterpret_cast<MaskType*>(out_col_valid);
+
+    // only one thread writes the mask
+    if (0 == threadIdx.x % warpSize)
+      out_mask32[out_mask_location] = result_mask;
+
+    // store the output value, one per thread
+    if (output_is_valid) {
+      if (average)
+        // special case for COUNT aggregator
+        out_col[i] = val / count;
+      else
+        out_col[i] = val;
+    }
+
+    // process next element 
+    i += stride;
+    active_threads = __ballot_sync(active_threads, i < nrows);
+  }
+}
+
+struct launch_kernel
+{
+  /**
+   * @brief Uses SFINAE to instantiate only for arithmetic types
+   */
+  template<typename ColumnType, template <typename AggType> class agg_op, bool average, class... TArgs,
+	   typename std::enable_if_t<std::is_arithmetic<ColumnType>::value, std::nullptr_t> = nullptr> 
+  void dispatch_aggregation_type(gdf_size_type nrows, TArgs... FArgs)
+  {
+    PUSH_RANGE("CUDF_ROLLING", GDF_ORANGE);
+
+    gdf_size_type block = 256;
+    gdf_size_type grid = (nrows + block-1) / block;
+
+    gpu_rolling<ColumnType, agg_op, average><<<grid, block>>>(nrows, FArgs...);
+
+    cudaDeviceSynchronize();
+    CUDA_CHECK_LAST();
+
+    POP_RANGE();
+  }
+
+  /**
+   * @brief If we cannot perform aggregation on this type then throw an error
+   */
+  template<typename ColumnType, template <typename AggType> class agg_op, bool average, class... TArgs,
+	   typename std::enable_if_t<!std::is_arithmetic<ColumnType>::value, std::nullptr_t> = nullptr> 
+  void dispatch_aggregation_type(gdf_size_type nrows, TArgs... FArgs)
+  {
+    CUDF_FAIL("Unsupported column type. Only arithmetic types are supported in rolling windows");
+  }
+
+  /**
+   * @brief Helper function for gdf_rolling. Deduces the type of the
+   * aggregation column and type and calls another function to invoke the
+   * rolling window kernel.
+   */
+  template <typename ColumnType>
+  void operator()(gdf_size_type nrows,
```

Comment:
I suggest adding a `stream` parameter that is defaulted to `0`. This will make it easier when we eventually add stream support to libcudf APIs if all of the underlying machinery already has stream arguments.

Reply:
Added in 4024dff